src/strategies/offer_maker/market_making/Kandel/abstract/CoreKandel.sol

```diff
@@ -0,0 +1,277 @@
+    }
+  }
+
+  ///@notice retracts and deprovisions offers of the distribution interval `[from, to[`
+  ///@param from the start index
+  ///@param to the end index
+  ///@dev this simply provisions this contract's balance on Mangrove.
+  ///@dev use in conjunction of `withdrawFromMangrove` if the user wishes to redeem the available WEIs
+  function retractOffers(uint from, uint to) external onlyAdmin {
+    uint collected;
+    for (uint index = from; index < to; index++) {
+      (IERC20 outbound_tkn, IERC20 inbound_tkn) = _tokenPairOfOrderType(OrderType.Ask);
+      collected +=
+        MGV.retractOffer(address(outbound_tkn), address(inbound_tkn), offerIdOfIndex(OrderType.Ask, index), true);
+      (outbound_tkn, inbound_tkn) = _tokenPairOfOrderType(OrderType.Bid);
+      collected +=
+        MGV.retractOffer(address(outbound_tkn), address(inbound_tkn), offerIdOfIndex(OrderType.Bid, index), true);
+    }
+  }
+
+  ///@notice takes care of reposting residual offer in case of a partial fill and logging potential issues.
+  ///@param order a recap of the taker order
+  ///@param makerData generated during `makerExecute` so as to log it if necessary
+  ///@return notPublished the amount of liquidity that failed to be published on mangrove
+  function _handleResidual(MgvLib.SingleOrder calldata order, bytes32 makerData) internal returns (uint notPublished) {
+    bytes32 repostStatus = super.__posthookSuccess__(order, makerData);
+    // Offer failed to repost for bad reason, logging the incident
+    if (repostStatus == "posthook/filled" || repostStatus == REPOST_SUCCESS) {
+      return 0;
+    }
+    if (repostStatus == "mgv/writeOffer/density/tooLow") {
+      return __residualGives__(order);
+    } else {
+      emit LogIncident(
+        MGV, IERC20(order.outbound_tkn), IERC20(order.inbound_tkn), order.offerId, makerData, repostStatus
+        );
+      return __residualGives__(order);
+    }
+  }
+
+  ///@notice repost residual offer and dual offer according to transport logic
+  ///@inheritdoc MangroveOffer
+  function __posthookSuccess__(MgvLib.SingleOrder calldata order, bytes32 makerData)
+    internal
+    override
+    returns (bytes32)
+  {
+    OrderType ba = _orderTypeOfOutbound(IERC20(order.outbound_tkn));
+    // adds any unpublished liquidity to pending[Base/Quote]
+    setPending(ba, _handleResidual(order, makerData));
+    // preparing arguments for the dual maker order
+    (OrderType dualBa, uint dualIndex, OfferArgs memory args) = _transportLogic(ba, order);
+    return _populateIndex(dualBa, dualIndex, args);
+  }
+
+  ///@notice In case an offer failed to deliver, promised liquidity becomes pending, but offer is not reposted.
+  ///@inheritdoc MangroveOffer
+  function __posthookFallback__(MgvLib.SingleOrder calldata order, MgvLib.OrderResult calldata)
+    internal
+    override
+    returns (bytes32)
+  {
+    OrderType ba = _orderTypeOfOutbound(IERC20(order.outbound_tkn));
+    setPending(ba, order.offer.gives());
```

Comment:
I wonder what is best to do here - because why did it fail? if we failed to deliver we will most likely also fail in the future? But yes this solution is probably as good as any.

Reply:
An natural invariant is that the sum of all offer gives + pending is equal to the balance of the contract's reserve (in base and quote). If this is no the case at some point, we will now how much is missing to restore the strat.